File: tests/fortran/init_test.f90
program init_test

#ifdef ELA_USE_MPI
use mpi
#endif

implicit none
integer, parameter :: N(3)=(/10,12,14/)
integer, parameter :: NN=2
integer, parameter :: pad(6)=(/0,0,0,0,0,0/)

integer, dimension(N(1)+pad(1)+pad(2),N(2)+pad(3)+pad(4),N(3)+pad(5)+pad(6)) :: labels
real(8), dimension(N(1)+pad(1)+pad(2),N(2)+pad(3)+pad(4),N(3)+pad(5)+pad(6)) :: vol
integer :: i,j,k,ierr,l

#ifdef ELA_USE_MPI
call MPI_Init(ierr)
if(ierr.ne.MPI_SUCCESS) error stop
#endif

! Initialize ELA
#ifdef ELA_USE_MPI
    call ELA_Init(N,pad,NN,SetupMPI())
#else
    call ELA_Init(N,pad,NN)
#endif

do concurrent(i=1:N(1),j=1:N(2),k=1:N(3))
    labels(i+pad(1),j+pad(2),k+pad(3))=i+j*N(1)+k*N(1)*N(2)
    vol   (i+pad(1),j+pad(2),k+pad(3))=1.0
end do

call ELA_InitLabels(vol,1,labels)

do i=1,N(1)
do j=1,N(2)
do k=1,N(3)
    call ELA_GetLabel(i,j,k,1,l)
    if(l.ne.(i+j*N(1)+k*N(1)*N(2))) error stop 'ELA_GetLabel()'
end do
end do
end do

call ELA_DeInit()

#ifdef ELA_USE_MPI
call MPI_Finalize(ierr)
if(ierr.ne.MPI_SUCCESS) error stop
#endif

contains

#ifdef ELA_USE_MPI
function SetupMPI() result(comm)
    implicit none
    integer :: comm, ierr
    integer, parameter :: ndims=3
    integer, parameter :: dims(3)=(/2,2,1/)
    logical, parameter :: periods(3)=(/.true.,.false.,.true./)

    call MPI_Cart_Create(MPI_COMM_WORLD,ndims,dims,periods,.false.,comm,ierr)
    if(ierr.ne.MPI_SUCCESS) error stop
end function
#endif

end program init_test

File: tests/cpp/init_test.cpp
#include <ELA.h>
#include <algorithm>
#include <gtest/gtest.h>

constexpr int N[3] = {10, 12, 14};

constexpr int NI = N[0];
constexpr int NJ = N[1];
constexpr int NK = N[2];
constexpr int NN = 2;

constexpr int DummyFunction(int i, int j, int k)
{
    return i + j * NI + k * NI * NJ;
    // return i+j*NI+k*NI*NJ;
}
int* newDummyArray()
{
    int* out = new int[NI * NJ * NK];

    int* itr = out;
#ifdef F_STYLE
    for (int k = 0; k < NK; k++) {
        for (int j = 0; j < NJ; j++) {
            for (int i = 0; i < NI; i++) {
#else
    for (int i = 0; i < NI; i++) {
        for (int j = 0; j < NJ; j++) {
            for (int k = 0; k < NK; k++) {
#endif
                *itr = DummyFunction(i, j, k);
                itr++;
            }
        }
    };

    return out;
};

TEST(ELA, Init)
{
    const int pad[6] = {0};

    ELA_Init(N, pad, NN);

    // Should get a label of zero if we try to call
    // ELA_GetLabel() before ELA_InitLabels()
    ASSERT_EQ(ELA_GetLabel(2, 3, 4, 0), 0);

    int* labels = newDummyArray();
    double* vol = new double[NI * NJ * NK];
    std::fill_n(vol, NI * NJ * NK, 1.0);

    ELA_InitLabels(vol, 0, labels);

    for (int k = 0; k < NK; k++) {
        for (int j = 0; j < NJ; j++) {
            for (int i = 0; i < NI; i++) {
                ASSERT_EQ(ELA_GetLabel(i, j, k, 0), DummyFunction(i, j, k));

                // unassigned labels are empty, should return zero
                ASSERT_EQ(ELA_GetLabel(i, j, k, 1), 0);
            }
        }
    }

    ELA_DeInit();
    delete[] labels;
    delete[] vol;
}